Mathlib/Topology/ApproximateUnit.lean

```diff
@@ -0,0 +1,148 @@
+- In a C⋆-algebra, the collection of sections `fun a ↦ {x | a ≤ x} ∩ ball 0 1`, where `a`
+  ranges over the positive elements of norm strictly less than 1, is the filter basis for an
+  approximate unit.
+
+## Main declarations
+
++ `Filter.mulLeftRightTendsto`: the largest filter `l` such that `Tendsto (· * m) l (𝓝 m)`
+  and `Tendsto (m * ·) l (𝓝 m)` for all `m`.
++ `ApproximateUnit`: a filter smaller than `Filter.mulLeftRightTendsto` with a basis all of whose
+  elements are contained in a given bounded set.
+
+## Implementation details
+
+Unlike most areas of the library, it is actually not the *filter* itself that is important, but
+rather the filter basis. Indeed, the only property relevant to the filter is that it is smaller
+than `Filter.mulLeftRightTendsto`. Moreover, in the case of C⋆-algebras, one always uses properties
+of the filter basis (e.g., that the sets are contained in the intersection of the positive cone
+and the unit ball). For this reason, we define `ApproximateUnit` as a structure extending
+both `Filter` and `Filter.HasBasis`, and we provide a convenience constructor for `Filter.IsBasis`.
+
+-/
+
+open Filter Topology Bornology
+
+section Def
+
+variable {α ι : Type*} [TopologicalSpace α] [Mul α]
+
+namespace Filter
+
+/-- The largest filter `l` such that `Tendsto (· * m) l (𝓝 m)` and `Tendsto (m * ·) l (𝓝 m)` -/
+@[to_additive "The largest filter `l` such that `Tendsto (· + m) l (𝓝 m)` and
+`Tendsto (m + ·) l (𝓝 m)`"]
+def mulLeftRightTendsto : Filter α :=
+  sSup { l : Filter α | ∀ (m : α), Tendsto (· * m) l (𝓝 m) ∧ Tendsto (m * ·) l (𝓝 m) }
+
+@[to_additive]
+lemma mulLeftRightTendsto.tendsto_mul_left (m : α) :
+    Tendsto (· * m) mulLeftRightTendsto (𝓝 m) := by
+  simp_rw [mulLeftRightTendsto, tendsto_def, mem_sSup]
+  aesop
+
+@[to_additive]
+lemma mulLeftRightTendsto.tendsto_mul_right (m : α) :
+    Tendsto (m * ·) mulLeftRightTendsto (𝓝 m) := by
+  simp_rw [mulLeftRightTendsto, tendsto_def, mem_sSup]
+  aesop
+
+@[to_additive]
+lemma mulLeftRightTendsto.le_iff {l : Filter α} :
+    l ≤ mulLeftRightTendsto ↔ ∀ m, Tendsto (· * m) l (𝓝 m) ∧ Tendsto (m * ·) l (𝓝 m) :=
+  ⟨fun h m ↦ ⟨(tendsto_mul_left m).mono_left h, (tendsto_mul_right m).mono_left h⟩,
+    fun h ↦ le_sSup h⟩
+
+end Filter
+
+variable [Bornology α]
+
+/-- An *approximate unit* is a filter smaller than `mulLeftRightTendsto` along with a basis
+consisting of sets contained in a given bounded set. The former condition is equivalent to the
+condition that for all `m`, multiplication on the left (or right) tends to `m` along the filter.
+
+Note: this extends `Filter` and `Filter.HasBasis` rather than taking the filter as an argument.
+This is because, in many cases, the only interesting property of the filter is that it is smaller
+than `mulLeftRightTendsto` and we may only be interested in the filter basis itself. -/
+structure ApproximateUnit (p : ι → Prop) (s : ι → Set α)
+    extends Filter α, toFilter.HasBasis p s where
```

Comment:
Why not define a predicate on `Filter`s instead? You can ask for 2 `Tendsto`s and `Disjoint l (cobounded α)`.

Reply:
Ah, that's a nice approach. I didn't think of the `Disjoint` spelling, and I'm also looking ahead to C*-algebras, where you really do want a basis with certain properties.

But for this definition I'll make the change to go basis-free.